### PR TITLE
Support multi-segment wells

### PR DESCRIPTION
**Pull request by atgeirr**
This PR adds a new derived model that supports multi-segment wells.

It is a very large PR, so it will be demanding to review. In the interest of saving time and effort we have tried to explain some choices we have made here, as well as provide some guidance as to what we think should be looked at.

The PR makes only limited changes to the existing `BlackoilModelBase` class, the new functionality is accessed through the new `BlackoilMultiSegmentModel` class.  A new main program, `flow_multisegment.cpp`, has been added that uses the new model. This has been done to minimise disruption to the regular version of flow. The only difference in the main programs is the simulator class used, so later refactoring of this should be trivial.

In the longer term we may want to refactor the whole well subsystem. The current effort may provide a starting point to such a refactoring, in that it uses the `Wells` C-struct (and thereby also the `WellsManager`) only in a minimal way. We do not consider the present proposal the end of the discussion of well APIs however!

There are some known limitations of this PR:
 - We have not yet made sure the new model will work in a parallel setting. This should be part of the discussion later however, when we refactor the well subsystem.
 - The VFP handling is disabled for the new model class.

In reviewing this, the most critical part is how existing classes are changed,
since that bears the risk of breaking existing functionality. The `BlackoilModelBase` model has
changed as follows:
 - The `WellOps` helper class now contain a vector `well_cells`. This simplifies code by eliminating repeated creation of this object in various functions.
 - The `assemble()` function has been simplified, and the extraction of the perforation cell properties has been moved to its own function, `extractWellPerfProperties()`.
 - Made a few functions `const` that should have been from the start.
 - Added a method `numWellVars()`, used in `updateState()`, enabling the derived model to use the base implementation of `updateState()`.
 - Made many method calls use asImpl(), to support subclass usage of base class implementations.
 - The `convergenceReduction()` method no longer requires the argument containing the number of wells.

The new classes are:
 - `WellStateMultiSegment`
 - `WellMultiSegment`
 - `SimulatorFullyImplicitBlackoilMultiSegment`
 - `BlackoilMultiSegmentModel`

`WellStateMultiSegment`: this class extends the existing well state class used by the base model with
data for the segments (pressures, fluxes) as well as a new well map to preserve segment information
across timesteps (since segments can be added to a well for example), the existing well map would
be insufficient for this. There is some duplication of functionality in the `init()` method here, however
designing one's way around that is better done later in my opinion, when we decide to remove the
`Wells` C struct from these codes completely, since the reason for duplication is precisely that the
method does not use that struct as input.

`WellMultiSegment`: this class provides an API for accessing all needed information for a multi-segment or regular well (regular wells can be thought of as having a single segment). It is (mostly) initialized from an opm-parser `Well` object, and for the most part does simple data storage. However it also contain some discrete operators (stored as sparse matrices) to map for example segment -> perforations or vice versa (for that single well).

`SimulatorFullyImplicitBlackoilMultiSegment`: the main difference from the plain `SimulatorBase` class is that the model requires a vector of (shared pointers to) `WellMultiSegment` well objects instead of the `Wells` struct. This means that it must implement its own version of `run()`, which mostly duplicates that of `SimulatorBase`. Again, this is a change that may foreshadow a well refactoring, and getting rid of this duplication is not worth it at this time in my opinion.

`BlackoilMultiSegmentModel`: this class implements the multi-segment well model. It is derived from
`BlackoilModelBase`, and makes the following modifications:
 - Add various data members for segment densities, depth differences etc.
 - Add a `std::vector<WellMultiSegment>`, replaces usage of `Wells`.
 - Add an instance of (new, private class) `MultiSegmentWellOps`, that is a richer set of discrete well operations than that of the base class.
 - Override the following methods with significant changes: constructor, `prepareStep()`, `numWellVars()`, `variableWellStateInitials()`, `variableStateExtractWellsVars()`, `computeWellConnectionPressures()`, `computeWellFlux()`, `updatePerfPhaseRatesAndPressures()`, `addWellFluxEq()`, `updateWellControls()`, `addWellControlEq()`, `updateWellState()`.
 - Override the following methods with small changes (usually calling the base class implementation): `makeConstantState()`, `assemble()`, `solveWellEq()`.
 - New important methods: `computeSegmentFluidProperties()`, `computeSegmentPressuresDelta()`
Some of the methods listed with significant changes do not add features, but essentially reimplement the base feature without using the Wells struct since we have to do without it here. This applies to for example `updateWellControls()`.
Basically, the new class modifies everything that has anything to do with wells, and leaves the rest unchanged.

### CODE REVIEW

**Comment by joakim-hove** [comment]
160 commits, nearly 3700 lines of new code and not a single line of test code - in my book that is just a immediate close and forget.

**Comment by atgeirr** [comment]
> not a single line of test code

There are integration tests, they are not part of this PR though.

**Comment by joakim-hove** [comment]
> they are not part of this PR though.

Well - not much help to the reviewer of this PR then. I have really "seen the light" when it comes to testing and test driven development and will continue to nag about this until I am kicked out of the project. There are three major benefits to be gained from test driven driven development:

1. A comprehensive test suite provides a safety net which make refactoring and new developments faster and safer.
2. The tests are the best possible documentation there is. Wonder how to instantiate a certain class - just go look at the test; in contrast to documentation the tests do not go stale.
3. *Testable code is good code*; even if you never actually run the tests(?) writing the code the in a *testable fashion* helps in creating loosely coupled code with well defined interfaces.

It is my impression that many developer who have not (yet) taken the blue pill focus mainly on point 1) - and since they generally write perfect bug free code they dismiss the whole idea of test driven development as a joke which does not apply to Real Programmers&trade; "*Come on write the test first - what are you smoking??*"  

In my opinon 2) and especially 3) are as important as 1); the opm codebase is quite opaque to new developers, taking the testing seriously and gaining some of the benefits under 2) and 3) would lower the bar for new developers.

**Comment by GitPaean** [comment]
To add something about the verification running. The regular wells are converted to be multi-segment wells with only one segment each. The code has been verified with SPE1, SPE9, Norne,  and examples with multi-segment wells written by our-self.

**Comment by joakim-hove** [comment]
Although I am obviously a rabid proponent of testing it was not my intention to derail the discussion in this PR - sorry. Further discussions of testing should be continued elsewhere.

**Comment by dr-robertk** [comment]
@GitPaean: It's good that you tested with the cases in opm-data. Nevertherless, this needs to be tested in parallel runs. I'll wait with the review until this is done.

**Comment by GitPaean** [comment]
I have never run flow under parallel mode. What is the current status? Can we run flow in parallel through SPE1, SPE3, SPE9, Norne?

**Comment by dr-robertk** [comment]
@atgeirr should know. I'm currently also not up to date on that one.

**Comment by atgeirr** [comment]
I would like not to add another (flow_multisegment_mpi?) program, but for testing purposes we could do this, ok.

**Comment by dr-robertk** [comment]
@atgeirr and @GitPaean: I have a couple of questions. 
1) Why yet another flow_something.cpp? Wouldn't this be the right time to stop this?
2) From the comments I read above and the answers I got I assume that this segmented wells will not function in parallel. So business as usual then? Implement some new features and then let somebody else fix the real problems arising from the need of parallelization. 
3) To review this will take some time. Just reading the explanations above take long..... 
4) Above it's claimed that the old wells are just Multisegment wells with one segment. Does that work for all the test cases? And if, we can come back to 2) immediately. 
5) There will be more questions.

**Comment by alfbr** [comment]
Just a couple of brief comments. It definitely is a good time to stop with the flow_something. However, this patch series is very intrusive, and not critical (yet), so it may need some maturation before replacing the current code. Ensuring that it plays nicely with mph is certainly one of these aspects. Another aspect is that the current PR adds significantly to the computational time of Norne model even though Norne has no multi-segment wells. Yet another factor is that this implementation serves as a starting point for restructuring the wells manager. Lastly, the patch set is already huge. In conclusion, it may be wise to keep it as a separate binary until at least the first review round is done. I do hope that all issues are ironed out though, and that this implementation will replace the current code.

As for mpi, I agree that from now on comitters should ensure that their code behaves nicely with parallel execution. We need a Jenkins job for this (as well as the latest work from Markus).

**Comment by dr-robertk** [comment]
@alfbr: I agree at that a PR that huge because of that it very difficult to forecast what the problems will be when merged directly with flow. 

However, I haven't run any test yet but if the segmented well code does not give the same performance for Norne, that I think that is a bit suspicious and definitely worth an investigation.

**Comment by atgeirr** [comment]
> 1) Why yet another flow_something.cpp? Wouldn't this be the right time to stop this?

This is my plan: make opm-polymer part of opm-autodiff, then unify all the flow variants by creating a template class (or using the Simulator* classes as a starting point) where model and grid are template arguments, so that the only differences between the flow_somethings would be the template arguments. The end result of that would still be several different flow_variants, but without code duplication. I tried to come up with a good inheritance-based design for the models when I refactored it earlier, but failed. So for now we have template-based polymorphism instead, which naturally leads to multiple programs if you want to make available all the different model variants. We can later combine them all into a single program and choose what to call based on input, but the essence would still be compile-time polymorphism.

A different approach would be to always add features to the base model, but I do not want to make a single monolithic model with all the functionality.

> 2) From the comments I read above and the answers I got I assume that this segmented wells will not function in parallel. So business as usual then? Implement some new features and then let somebody else fix the real problems arising from the need of parallelization. 

In some sense, yes. I do not know very well the details required to do this properly, and would need help from you or Markus to do it right. I think we should try to separate more clearly in which parts of the code one can "think serially" and which parts you cannot. Right now, I assume we're doing something wrong since we skip the WellsManager treatment, but I am not sure exactly what would be needed.

> 3) To review this will take some time. Just reading the explanations above take long..... 

Sure, I hope you consider the explanations a good thing though, I wrote it to make it easier to deal with the PR!

> 4) Above it's claimed that the old wells are just Multisegment wells with one segment. Does that work for all the test cases? And if, we can come back to 2) immediately. 

We've tested it with SPE9 (identical results) and also run it on Norne. For that case, minor changes in the computations (ordering perhaps?) lead eventually to time-stepping differences, as usual this makes it harder to compare directly the output. I did not understand what you mean coming back to 2?

**Comment by atgeirr** [comment]
> this needs to be tested in parallel runs

I tried this (with SPE9) and unsurprisingly it fails. I get exception thrown from line 197 of WellMultiSegment.cpp since it does not find a well (and for a mswell case line 83 would trigger). I would appreciate any help in getting this right. Note that this part might still perhaps benefit from the parallel work done in the WellsManager, since the Wells struct is used for initialization purposes (but not in the model class as pointed out earlier).

**Comment by GitPaean** [comment]
Added a `flow_multisegment_mpi` for the parallel running. It can run through SPE9_CP now.

**Comment by atgeirr** [comment]
@dr-robertk, it looks to us like the init modification is sufficient to make this work in parallel (it does however lean on the work done in WellsManager to handle the "should I deal with this well" question). I am relieved that it seems to be less complicated than I feared!

**Comment by dr-robertk** [comment]
Great. Good work. I will look into the code now. It might take a while.

**Comment by dr-robertk** [comment]
@atgeirr & @GitPaean : flow_multisegment does not compile. Please fix this and I will finish my review tomorrow.

**Comment by dr-robertk** [comment]
WellMultiSegment: the typedefs for V and M should be renamed to Vector and Matrix. I think M and V are hard to read and hard to grep.

**Comment by atgeirr** [comment]
The fix for the compile failure is simple: on line 62 of SimulatorFullyImplicitBlackoilMultiSegment.hpp the `const` must be removed from the DerivedGeology parameter. @GitPaean, you must actually push the fix since the PR is against your repo.

**Comment by GitPaean** [comment]
fixed compilation error. Thanks.

**Comment by GitPaean** [comment]
fixed the type naming for `Matrix`. The `V` was deleted since it turned out not to be used.

**Comment by dr-robertk** [comment]
Thanks, the code compiles now. 

I have investigated the behavior for SPE9 and NORNE. 
When using flow_multisegment the code runs between 5% (SPE9) and 7% (NORNE) slower. This is a bit surprising since the single wells should be multisegment wells with one segment. So (after this PR has been merged) there should be some performance improvement addressing this. 
I will send the well comparison via email. The comparison seems fine. 

Other than that the code looks like a Fortran programmer was forced to code in C++. For that reason I'm not able to judge whether there are major problems with this piece or not. It's just to much, and coding style or not, I find it close to unreadable (in the given short time I have to review this). 

Also, as blattms had pointed out earlier, there is little benefit from setting up matrices and then do one multiplication (such as in BlackoilMultiSegmentModel_impl.hpp MultiSegmentWellOps::
MultiSegmentWellOps, line 111 and following. 

To sum up, I will merge this tomorrow, but I strongly suggest to address @joakim-hove point, that thorough testing needs to be added. Currently there is only one test. I'm not sure if this covers all issues. The second point is performance, as pointed out above.

**Comment by atgeirr** [comment]
> So (after this PR has been merged) there should be some performance improvement addressing this

For this, I think we should direct our work towards first agreeing on a how to refactor the whole well subsystem. When we are done, I'd like to be able to easily specify new well models and still have them solved fully implicitly for example. There should in the end be no performance loss, it should also be possible to be faster than the current regular well implementation. The first step must be to discuss what this should look like, what the interfaces should be etc. Otherwise I fear that investing lots of time speeding up either of the current well implementations now is a risk (potentially wasting effort), since they may be superseded.

**Comment by dr-robertk** [comment]
@atgeirr: I agree, that the interface design decision has to be made first. Provisional arrangement usually stay the longest. I'll merge this now.
